File: java/datastructure/array/leetcode/SingleNumber.java
package datastructure.array.leetcode;

import java.util.HashSet;
import java.util.Set;

/**
 * @author roseduan
 * @time 2020/11/16 6:36 下午
 * @description 只出现一次的数字
 */
public class SingleNumber {

    /**
     * 使用一个set
     */
    public int singleNumber1(int[] nums) {
        Set<Integer> set = new HashSet<>();
        for (int n : nums) {
            if (set.contains(n)) {
                set.remove(n);
            } else {
                set.add(n);
            }
        }
        return set.iterator().next();
    }

    /**
     * 位运算
     */
    public int singleNumber2(int[] nums) {
        int res = 0;
        for (int n : nums) {
            res ^= n;
        }
        return res;
    }
}

File: cpp/array/single_number.cpp
#include "vector"
#include "set"
#include "iostream"
using namespace std;

// 只出现一次的数字

class Solution {

public:
    // 使用一个哈希表
    int singleNumber1(vector<int>& nums) {
        set<int> s;
        for (int i : nums) {
            if (s.count(i)) {
                s.erase(i);
            } else {
                s.insert(i);
            }
        }

        int res = 0;
        for (int i : s) {
            res = i;
        }
        return res;
    }

    // 位运算
    int singleNumber2(vector<int>& nums) {
        int c = 0;
        for (int i : nums) {
            c = c ^ i;
        }
        return c;
    }
};
